### PR TITLE
Added State to bruin run 

### PR DESCRIPTION
**Pull request by y-bruin**
State Example 
```
       │ File: logs/runs/2024_12_19_19_15_17.json
───────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ {
   2   │   "parameters": {
   3   │     "endDate": "2024-12-18T23:59:59Z",
   4   │     "excludeTag": "",
   5   │     "fullRefresh": "false",
   6   │     "only": "",
   7   │     "pipeline": "bruin-init",
   8   │     "runID": "2024_12_19_19_15_17",
   9   │     "startDate": "2024-12-18T00:00:00Z",
  10   │     "tag": "",
  11   │     "useUv": "false"
  12   │   },
  13   │   "metadata": {
  14   │     "version": "dev",
  15   │     "os": "darwin"
  16   │   },
  17   │   "state": [
  18   │     {
  19   │       "name": "example4",
  20   │       "status": "failed"
  21   │     },
  22   │     {
  23   │       "name": "example",
  24   │       "status": "failed"
  25   │     },
  26   │     {
  27   │       "name": "example2",
  28   │       "status": "succeeded"
  29   │     },
  30   │     {
  31   │       "name": "example3",
  32   │       "status": "succeeded"
  33   │     }
  34   │   ],
  35   │   "version": "1.0.0",
  36   │   "timestamp": "2024-12-19T19:15:17.641044+05:30",
  37   │   "run_id": "2024_12_19_19_15_17",
  38   │   "compatibility_hash": "4a8e865ced3e067052690f1926f1d2de3e58ad77bbbdf9e05cba6b4285dc9b29"
  39   │ }
```

### CODE REVIEW

File: pkg/state/state.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+			OS:      runtime.GOOS,
+		},
+		State:     map[string][]*SchedulerState{},
+		Version:   "1.0.0",
+		TimeStamp: time.Time{},
+		RunID:     runid,
+	}
+}
+
+func (s *State) InitState(foundPipelines []pipeline.Pipeline) error {
```
Comment (karakanb) [inline pkg/state/state.go:62]: I wouldn't have done it this way, this means you'll have to update the state step by step at every execution from the scheduler, which is duplicate work. the scheduler already stores this state, I would couple it with that, and just create a struct that can be saved from the scheduler.

**Review by karakanb — COMMENTED** [review]
I would suggest coupling this implementation more with the scheduler, just a function that will take the scheduler and convert its internal state to a file.

File: pkg/state/state.go [hunk 2]
```diff
@@ -11,32 +24,105 @@ const (
...
 }
 
 type Metadata struct {
-	Version string
-	OS      string
+	Version string `json:"version"`
+	OS      string `json:"os"`
 }
 
 type SchedulerState struct {
```
[the remark this answers is not in the document]
Comment (y-bruin) [reply pkg/state/state.go:36]: yes

File: pkg/state/state.go [hunk 3]
```diff
@@ -11,32 +24,105 @@ const (
...
+	timestamp := time.Now().Format(time.RFC3339)
+	filePath := fmt.Sprintf("%s/%s_%s.json", stateFileDir, timestamp, runid)
+	file, err := os.Create(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create file: %v", err)
+	}
+	return file, nil
+}
+
+func NewState(runid string, parameters map[string]string, stateFileDir string, foundPipelines []pipeline.Pipeline) *State {
```
[the remark this answers is not in the document]
Comment (y-bruin) [reply pkg/state/state.go]: i will, Just reverted unwanted code for now

File: pkg/state/state.go [hunk 4]
```diff
@@ -46,20 +44,23 @@ type SchedulerState struct {
...
-	Id        string              `json:"id"`
+	ID        string              `json:"id"`
 }
 
 type Status string
 
 func createStateFile(runid string, stateFileDir string) (*os.File, error) {
 	if _, err := os.Stat(stateFileDir); os.IsNotExist(err) {
-		os.Mkdir(stateFileDir, 0755)
+		err := os.Mkdir(stateFileDir, 0o755)
```
[the remark this answers is not in the document]
Comment (y-bruin) [reply pkg/state/state.go]: yes but for now i just want to commit the contract

**Review by albertobruin — APPROVED** [review]

**Pull request merged**